### PR TITLE
refactor: remove git CLI dependency from log / diff code paths

### LINKED ISSUE #69: refactor: remove `git` CLI dependency from log / diff code paths
## Background

rvpm is intended to be fully gix-based with no runtime dependency on the `git` CLI binary. Two `git` subprocess call sites crept back in when `rvpm log` was added (PR #52, commit `fbd2934`) and are still in the tree on v3.18.0. Both have resilient fallbacks, so the binary doesn't fail without `git` installed — but the "no-git-required" invariant is silently broken.

### 1. `src/git.rs:300` — `doc_files_changed()`

```rust
// git diff --name-only <from>..<to> -- README* CHANGELOG* doc/
```

- Invoked from `build_change()` whenever `Repo::sync()` / `update()` actually advances HEAD, i.e. **on every successful pull that produces a change**.
- Populates `GitChange::doc_files_changed`, which is persisted to `update_log.json` for later consumption by `rvpm log`.
- Resilience: git missing or exit != 0 → empty `Vec`, sync continues.

### 2. `src/main.rs:3871` — patch rendering in `run_log --diff`

```rust
// git diff <from>..<to> -- <file>
```

- Invoked only when the user passes `--diff` to `rvpm log`.
- Retrieves the actual unified diff text for README / CHANGELOG / doc files.
- Resilience: failure → skip that file's patch.

## Why the subprocess was chosen

`src/git.rs:298` has an honest comment: `// gix の diff API は複雑なため subprocess`. The gix diff pipeline (tree walk → blob diff → unified patch formatting) is lower-level than `std::process::Command::new("git")`. That tradeoff was deliberate but is now a papercut against the no-git-required positioning.

## User-visible impact

- On a system without `git`, `rvpm log` loses the `doc_files_changed` list entirely (first call site) and the `--diff` patch embedding is silently skipped (second call site). `rvpm sync` / `generate` / `update` themselves still work.
- `rvpm doctor` doesn't currently flag missing `git` as an issue, so users may be surprised to find `rvpm log` partially functional.

## Proposed remediation (pick one)

- **A. Document and stop.** Mark `git` as an *optional* runtime dep in the README and `rvpm doctor`. No code change. Cheapest, but keeps the invariant broken.
- **B. Full gix refactor.** Use `gix_diff`'s tree-walk + blob-diff APIs for the name-only list and an in-tree unified-diff formatter (or a small crate like `imara-diff` / `similar`) for the patch output. Restores the invariant. Larger diff, especially for patch formatting.
- **C. Phased: name-only first, patch later.** Replace `doc_files_changed` with `gix_diff` tree-walk (runs on every sync, so higher leverage for correctness). Leave `--diff` patch rendering on subprocess for a separate pass. Compromise between A and B.

Leaning toward **C** — the sync-path call site matters more (it runs silently on every pull and writes stale data to `update_log.json` when `git` is missing), and the patch-rendering call site is gated behind an explicit `--diff` flag so the degradation is at least visible.

## Related

- Introduced in #52 (commit `fbd2934`)
- Discovered during post-v3.18.0 audit (rvpm.nvim integration work)

### PR DESCRIPTION
**Pull request by yukimemi**
## Summary

Restores rvpm's "no `git` CLI required at runtime" invariant by replacing the two `git` subprocess invocations introduced in #52 with `gix`-native equivalents. Closes #69.

- `doc_files_changed` (sync path, runs on every successful pull) → `Repository::diff_tree_to_tree` + `is_doc_path` predicate equivalent to the previous `README* CHANGELOG* doc/` pathspec.
- `rvpm log --diff` patch rendering → `gix_diff::blob::UnifiedDiff` with `byte_lines` tokenization. Emits the standard `diff --git` / `index` / `--- a/...` / `+++ b/...` headers and unified hunks. Binary blobs are detected via the null-byte heuristic and rendered as a one-line `Binary files ... differ` notice.
- Both call sites share a `tree_changes` helper that swallows open / rev-parse / tree-peel failures into an empty `Vec` to preserve the existing best-effort resilience.
- `rvpm doctor` no longer flags a missing `git` as `Severity::Error`. It is now reported as `(optional)`.

The only remaining `git` invocation in the tree is in `git_cmd` inside the `git.rs` test module, which builds fixture repos with the CLI — this is dev-time only and not in the runtime path.

## Test plan

- [x] `cargo test` (719 passed, 1 ignored)
- [x] `cargo fmt --all -- --check`
- [x] `cargo clippy --all-targets -- -D warnings`
- [x] New tests cover `is_doc_path`, `doc_files_changed` (filter set + missing-repo resilience), `doc_file_patch` (modification, addition, no-match → `None`), and `check_tool_git` for the missing case.
- [ ] Manual smoke: `rvpm log --diff` against a real plugin upgrade with README changes.

## Summary by CodeRabbit

* **Bug Fixes**
  * Git is now marked as optional instead of required—missing git no longer produces an error or installation prompt.

* **Improvements**
  * Doc file diff generation improved through more efficient internal processing.

* **Documentation**
  * Added CLI maintenance checklist to ensure tool completions remain synchronized with core updates.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -5157,16 +5157,9 @@ async fn run_log(query: Option<String>, last: usize, full: bool, diff: bool) ->
...
-                    let output = tokio::process::Command::new("git")
-                        .arg("-C")
-                        .arg(&dst_path)
-                        .args(["diff", &format!("{}..{}", from, change.to), "--", file])
-                        .output()
-                        .await;
-                    if let Ok(out) = output
-                        && out.status.success()
+                    if let Some(patch) =
+                        crate::git::doc_file_patch(&dst_path, from, &change.to, file)
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/main.rs]: @gemini-code-assist Fixed in e26529b. Added `doc_file_patches` (batch API) so the repo is opened and the tree diff computed exactly once per plugin update — `run_log --diff` now iterates the precomputed map instead of re-doing the open/parse/diff per file.

File: src/git.rs [hunk 2]
```diff
@@ -373,44 +373,208 @@ fn split_subject_body(msg: &str) -> (&str, &str) {
...
+    use gix::object::tree::diff::ChangeDetached;
+
+    let repo = gix::open(dst).ok()?;
+    let from_id = repo.rev_parse_single(from).ok()?;
+    let to_id = repo.rev_parse_single(to).ok()?;
+    let from_tree = repo.find_commit(from_id).ok()?.tree().ok()?;
+    let to_tree = repo.find_commit(to_id).ok()?.tree().ok()?;
+    let changes = repo
+        .diff_tree_to_tree(Some(&from_tree), Some(&to_tree), None)
+        .ok()?;
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/git.rs]: @gemini-code-assist Extracted `open_and_diff` (commit e26529b) which performs repo open + rev parse + tree peel + tree diff once. Both `doc_files_changed` and the new `doc_file_patches` batch through it now.

File: src/git.rs [hunk 3]
```diff
@@ -373,44 +373,208 @@ fn split_subject_body(msg: &str) -> (&str, &str) {
...
+    match change {
+        ChangeDetached::Addition { location, .. }
+        | ChangeDetached::Deletion { location, .. }
+        | ChangeDetached::Modification { location, .. } => vec![location.to_string()],
+        ChangeDetached::Rewrite {
+            source_location,
+            location,
+            ..
+        } => vec![source_location.to_string(), location.to_string()],
+    }
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/git.rs]: @gemini-code-assist Good catch. Fixed in e26529b by explicitly disabling rewrite tracking — `open_and_diff` now passes `Options::default().with_rewrites(None)` to `diff_tree_to_tree`, so renames surface as separate Deletion + Addition entries (matching the prior `git diff --name-only` no-rename semantics). The `change_paths` helper is now `change_location` returning a single path; the defensive Rewrite arm is unreachable in normal flow but kept as a safety net.

File: src/git.rs [hunk 4]
```diff
@@ -373,44 +373,208 @@ fn split_subject_body(msg: &str) -> (&str, &str) {
...
+    let change = changes.into_iter().find(|c| match c {
+        ChangeDetached::Addition { location, .. }
+        | ChangeDetached::Deletion { location, .. }
+        | ChangeDetached::Modification { location, .. } => location.as_slice() == path_bytes,
+        ChangeDetached::Rewrite {
+            source_location,
+            location,
+            ..
+        } => location.as_slice() == path_bytes || source_location.as_slice() == path_bytes,
+    })?;
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/git.rs]: @gemini-code-assist Fixed in e26529b. With `with_rewrites(None)` the find closure now matches only `location` for all variants (Rewrite included). The Rewrite arm in the body match is unreachable in the normal path but kept defensively.

File: src/git.rs [hunk 5]
```diff
@@ -373,44 +373,208 @@ fn split_subject_body(msg: &str) -> (&str, &str) {
...
+            lines: &[(DiffLineKind, &[u8])],
+        ) -> std::io::Result<()> {
+            // HunkHeader implements Display as `@@ -A,B +C,D @@`.
+            self.0.push_str(&format!("{}\n", header));
+            for (kind, line) in lines {
+                self.0.push(kind.to_prefix());
+                self.0.push_str(&String::from_utf8_lossy(line));
+                if !line.ends_with(b"\n") {
+                    self.0.push('\n');
+                }
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/git.rs:590]: @gemini-code-assist Per the gix-diff 0.62 docs for [`byte_lines`](https://docs.rs/gix-diff/0.62.0/gix_diff/blob/sources/fn.byte_lines.html), the newline separator (`\r\n` or `\n`) **is** included in the emitted tokens — this is what lets the diff detect `\r\n` ↔ `\n` changes or a missing trailing newline on the last line. So `line.ends_with(b"\n")` is true for normal lines and only false for an unterminated final line; the conditional is there to keep the unified output well-formed in that edge case. I added a regression test in e26529b (`test_doc_file_patch_lines_are_separated`) that asserts each `-` / `+` line still carries its own `\n`, so this will not silently regress if gix changes.

File: src/git.rs [hunk 6]
```diff
@@ -373,44 +373,208 @@ fn split_subject_body(msg: &str) -> (&str, &str) {
...
+            id.to_string(),
+            "0000000".to_string(),
+        ),
+        ChangeDetached::Rewrite { source_id, id, .. } => (
+            read_blob(source_id).unwrap_or_default(),
+            read_blob(id).unwrap_or_default(),
+            source_id.to_string(),
+            id.to_string(),
+        ),
+    };
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/git.rs]: @coderabbitai Fixed in e26529b along both axes:

1. **Rewrite handling**: Disabled rewrite tracking entirely via `Options::default().with_rewrites(None)` in the new `open_and_diff` helper. Renames now surface as separate Deletion + Addition entries (matching the prior `git diff --name-only` no-rename behavior), so the duplicate-patch ambiguity goes away. The Rewrite match arm in `build_patch_for_path` is unreachable in normal flow but kept defensively.

2. **Blob fallback**: Replaced `read_blob(...).unwrap_or_default()` with `read_blob(...)?` so a missing blob short-circuits the patch (the path is simply absent from the returned `HashMap` rather than emitting an empty/misleading diff).

Re: testing the Rewrite branch — since rewrite tracking is now disabled at the diff-options level, there is no realistic in-tree way to exercise that arm. I added a binary-blob regression test (`test_doc_file_patch_handles_binary_blob`) and a path-not-in-diff test (`test_doc_file_patches_skips_paths_not_in_diff`) which together cover the most fragile branches that remain reachable.
